### PR TITLE
feat: encrypt proteus message via CoreCrypto FS-1554

### PR DESCRIPTION
**Pull request by johnxnguyen**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

In this PR we add support for encrypting proteus messages via CoreCrypto. Changes include:

- Refactoring the `GenericMessage ` encryption extension methods to inject an encryption function.
- For each encryption method, create two versions, one that injects an encryption function via CoreCrypto, the other (legacy) that injects the encryption function via Cryptobox.
- Checking the `DeveloperFlag.proteusViaCoreCrypto` flag and calling one of the two encryption method variants.

#### Test Coverage

- Split `ZMClientMessageTests_OTR` to have a legacy variant that runs tests with `DeveloperFlag.proteusViaCoreCrypto` off, and the other variant with the flag on
- Unit tests for the `ProteusService` encryption methods.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessageTests+OTR_Legacy.swift [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+// 
+
+import XCTest
+@testable import WireDataModel
+
+final class ClientMessageTests_OTR_Legacy: BaseZMClientMessageTests {}
+
+// MARK: - Payload creation
+extension ClientMessageTests_OTR_Legacy {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessageTests+OTR_Legacy.swift:25]: These tests are what was already existing, I just renamed it include the word "Legacy", we'll delete all these tests later when we remove Cryptobox.

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 2]
```diff
@@ -211,98 +211,238 @@ extension GenericMessage {
...
+            let proteusService = context.proteusService,
+            let selfClient = ZMUser.selfUser(in: context).selfClient(),
+            selfClient.remoteIdentifier != nil
+        else {
+            return nil
+        }
+
+        var messageData: Data?
+
+        // TODO: get core crypto file lock
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift:346]: what does this TODO mean? Don't we have file locks already with safe core crypto?

Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift:346]: It's not in this branch yet.

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 3]
```diff
@@ -403,50 +555,61 @@ extension GenericMessage {
...
         }
 
-        return nil
+        do {
+            let plainText = try serializedData()
+            let encryptedData = try encryptionFunction(sessionID, plainText)
+            guard let data = encryptedData else { return nil }
+            return Proteus_ClientEntry(withClient: client, data: data)
+        } catch {
+            print("error: \(String(describing: error))")
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]: Should we use log instead?

**Review by David-Henner — APPROVED** [review]

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]: Good catch!

Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]: Done

**Review by netbe — APPROVED** [review]

**Pull request merged**